packages/hasher/src/yarnWorkspaces.ts

```diff
@@ -0,0 +1,119 @@
+import * as path from "path";
+import * as fg from "fast-glob";
+
+import * as findWorkspaceRoot from "find-yarn-workspace-root";
+
+type PackageJsonWorkspaces = {
+  workspaces?:
+    | {
+        packages?: string[];
+        nohoist?: string[];
+      }
+    | string[];
+};
+
+export type WorkspaceInfo = { name: string; path: string }[];
+
+function getYarnWorkspaceRoot(cwd: string) {
+  const yarnWorkspacesRoot = findWorkspaceRoot(cwd);
+
+  if (!yarnWorkspacesRoot) {
+    throw new Error("Could not find yarn workspaces root");
+  }
+
+  return yarnWorkspacesRoot;
+}
+
+function getRootPackageJson(yarnWorkspacesRoot: string) {
+  const packageJson = require(path.join(yarnWorkspacesRoot, "package.json"));
+
+  if (!packageJson) {
+    throw new Error("Could not load package.json from workspaces root");
+  }
+
+  return packageJson;
+}
+
+function getPackages(packageJson: PackageJsonWorkspaces) {
+  const { workspaces } = packageJson;
+
+  if (workspaces && Array.isArray(workspaces)) {
+    return workspaces;
+  }
+
+  if (!workspaces || !workspaces.packages) {
+    throw new Error("Could not find a workspaces object in package.json");
+  }
+
+  return workspaces.packages;
+}
+
+function getPackagePaths(yarnWorkspacesRoot: string, packages: string[]) {
+  const packagePaths = packages.map(name =>
+    fg.sync(name, {
+      cwd: yarnWorkspacesRoot,
+      onlyDirectories: true,
+      absolute: true
+    })
+  );
+
+  return packagePaths.reduce((acc, cur) => {
+    return [...acc, ...cur];
+  });
+}
+
+export function getWorkspacePackageInfo(
+  workspacePaths?: string[]
+): WorkspaceInfo {
+  if (!workspacePaths) {
+    return [];
+  }
+
+  return workspacePaths
+    .map(workspacePath => {
```

Comment:
A map+filter is a reduce. If you use reduce, you don't need to do this type cast:
`const a = b.map(b => isValid(b) ? b.name : undefined).filter(Boolean)` could be coded in a cleaner fashion this way:
`const a = b.reduce((acc, curr) =>  isValid(curr) ? [...acc, curr.name] : acc, [])`

By using reduce you get correct type checking from TypeScript (casting disables type checking)

Reply:
Great insight. Thanks!!!